iseq.c

```diff
@@ -248,230 +245,155 @@ rb_iseq_each_value(const rb_iseq_t *iseq, iseq_value_itr_t * func, void *data)
+        rb_gc_mark_and_move(&body->variable.pc2branchindex);
+        rb_gc_mark_and_move(&body->variable.script_lines);
+        rb_gc_mark_and_move(&body->location.label);
+        rb_gc_mark_and_move(&body->location.base_label);
+        rb_gc_mark_and_move(&body->location.pathobj);
+        if (body->local_iseq) rb_gc_mark_and_move_ptr(&body->local_iseq);
+        if (body->parent_iseq) rb_gc_mark_and_move_ptr(&body->parent_iseq);
+        if (body->mandatory_only_iseq) rb_gc_mark_and_move_ptr(&body->mandatory_only_iseq);
 
         if (body->call_data) {
-            struct rb_call_data *cds = (struct rb_call_data *)body->call_data;
-            for (unsigned int i=0; i<body->ci_size; i++) {
-                const struct rb_callinfo *ci = cds[i].ci;
-                const struct rb_callcache *cc = cds[i].cc;
+            for (unsigned int i = 0; i < body->ci_size; i++) {
+                struct rb_call_data *cds = body->call_data;
 
-                if (vm_ci_markable(ci)) {
-                    rb_gc_mark_movable((VALUE)ci);
+                if (reference_updating) {
+                    if (!SPECIAL_CONST_P((VALUE)cds[i].ci)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].ci);
+                    }
+                }
+                else {
+                    if (vm_ci_markable(cds[i].ci)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].ci);
+                    }
                 }
 
-                if (cc) {
-                    VM_ASSERT((cc->flags & VM_CALLCACHE_ON_STACK) == 0);
+                if (cds[i].cc && (reference_updating || vm_cc_markable(cds[i].cc))) {
+                    VM_ASSERT(reference_updating || (cds[i].cc->flags & VM_CALLCACHE_ON_STACK) == 0);
 
-                    if (vm_cc_markable(cc)) {
-                        if (!vm_cc_invalidated_p(cc)) {
-                            rb_gc_mark_movable((VALUE)cc);
-                        }
-                        else {
-                            cds[i].cc = rb_vm_empty_cc();
-                        }
+                    if (reference_updating || !vm_cc_invalidated_p(cds[i].cc)) {
+                        rb_gc_mark_and_move_ptr(&cds[i].cc);
+                    }
+                    else {
+                        cds[i].cc = rb_vm_empty_cc();
                     }
                 }
             }
         }
 
         if (body->param.flags.has_kw && ISEQ_COMPILE_DATA(iseq) == NULL) {
             const struct rb_iseq_param_keyword *const keyword = body->param.keyword;
-            int i, j;
-
-            i = keyword->required_num;
 
-            for (j = 0; i < keyword->num; i++, j++) {
-                VALUE obj = keyword->default_values[j];
-                if (!SPECIAL_CONST_P(obj)) {
-                    rb_gc_mark_movable(obj);
+            for (int j = 0, i = keyword->required_num; i < keyword->num; i++, j++) {
+                if (!SPECIAL_CONST_P(keyword->default_values[j])) {
```

Comment:
The function body of `rb_gc_mark_and_move` starts with `if (RB_SPECIAL_CONST_P(*ptr)) return;`.  I think this `if` statement (and many similar ones) is unnecessary.

Reply:
👍 Good catch!